src/Voron/Data/CompactTrees/CompactTree.Dictionaries.cs

```diff
@@ -0,0 +1,264 @@
+//  - Cons: We may need larger dictionaries to accomodate better compression ratios. 
+//
+//  In Voron we use global scopes dictionaries because they have a cleaner implementation with less intrusive processes
+//  where it really matters (writing). 
+
+unsafe partial class CompactTree
+{
+    public struct FullDictionaryKeyScanner : IReadOnlySpanEnumerator
+    {
+        private readonly CompactTree _tree;
+        private Iterator _iterator;
+
+        public FullDictionaryKeyScanner(CompactTree tree)
+        {
+            _tree = tree;
+            _iterator = _tree.Iterate();
+            _iterator.Reset();
+        }
+
+        public bool MoveNext(out ReadOnlySpan<byte> result)
+        {
+            bool operationResult = _iterator.MoveNext(out Span<byte> resultSlice, out long _);
+            result = resultSlice;
+            return operationResult;
+        }
+
+        public void Reset()
+        {
+            _iterator = _tree.Iterate();
+            _iterator.Reset();
+        }
+    }
+
+    public struct RandomDictionaryKeyScanner : IReadOnlySpanEnumerator
+    {
+        private readonly CompactTree _tree;
+        private readonly int _samples;
+        private readonly int _seed;
+        private RandomIterator _iterator;
+
+        public RandomDictionaryKeyScanner(CompactTree tree, int samples, int seed = 0)
+        {
+            _tree = tree;
+            _seed = seed;
+            _samples = samples;
+            _iterator = _tree.RandomIterate(samples, seed: seed);
+            _iterator.Reset();
+        }
+
+        public bool MoveNext(out ReadOnlySpan<byte> result)
+        {
+            bool operationResult = _iterator.MoveNext(out Span<byte> resultSlice, out long _);
+            result = resultSlice;
+            return operationResult;
+        }
+
+        public void Reset()
+        {
+            _iterator = _tree.RandomIterate(_samples, seed: _seed);
+            _iterator.Reset();
+        }
+    }
+
+    public struct RandomBranchDictionaryKeyScanner : IReadOnlySpanEnumerator
+    {
+        private readonly long _samples;        
+        private readonly Random _generator;
+        private readonly CompactTree _tree;
+        private int _currentSample;
+        private int _currentIdx;
+        private IteratorCursorState _cursor;
+        private HashSet<long> _pagesSeen;
+
+        public RandomBranchDictionaryKeyScanner(CompactTree tree, long samples, int seed = 0)
+        {
+            _tree = tree;
+            _samples = samples;
+            _currentSample = 0;
+            _currentIdx = 0;
+            _generator = (seed == 0) ? new() : new(seed);
+            _pagesSeen = new HashSet<long>();
+
+            _cursor = new() { _stk = new CursorState[8], _pos = -1, _len = 0 };
+            _tree.PushPage(_tree._state.RootPage, ref _cursor);            
+        }
+
+        public bool MoveNext(out ReadOnlySpan<byte> key)
```

Comment:
I really like this approach, we are giving lots of keys (in order) from branches and creating good mix.

Reply:
I like it too, much better than fully randomized. Much faster are almost the same compression ratio.